client/domain/ReinforcementObjectFactory.js

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating new frontend instances of Reinforcement
+ * domain objects.
+ */
+
+tie.factory('ReinforcementObjectFactory', [
+  'ReinforcementBulletObjectFactory', function(
+      ReinforcementBulletObjectFactory) {
+    /**
+    * The reinforcement object contains the current state of reinforcement
+    * comments about a particular task in a question.
+    * Fields:
+    *  task: The corresponding task for the reinforcement comments.
+    *  passedTags: Dictionary containing tags (test sets) that the code
+    *              passed/failed.
+    *              Key: The tag
+    *              Value: Boolean indicating whether the code passed the tag.
+    *  pastFailedCases: The list of test cases that will be displayed in the
+    *                   reinforcement comments and the corresponding pass/fail
+    *                   status.
+    *                   Key: Stringified input of test case.
+    *                   Value: Boolean indicating where the code passed the
+    *                          test case.
+    */
+    var Reinforcement = function(task) {
+      this._task = task;
+      this._passedTags = {};
+      this._pastFailedCases = {};
+    };
+
+    // Instance methods.
+    Reinforcement.prototype.getBullets = function() {
+      var bullets = [];
+      for (var tag in this._passedTags) {
+        if (this._passedTags[tag]) {
+          bullets.push(
+            ReinforcementBulletObjectFactory.createPassedBullet(
+              "Handles " + tag));
+        } else {
+          bullets.push(
+            ReinforcementBulletObjectFactory.createFailedBullet(
+              "Fails " + tag));
+        }
+      }
+
+      for (var testCase in this._pastFailedCases) {
```

Comment:
var stringifiedTestCase

Recommend similar changes elsewhere throughout this file (and possibly others), if necessary, so that it's always clear whether we're dealing with a stringified version or the raw version.